File: lessons/l1_cache_hit_rate/l1_cache_hit_rate.go
package main

import (
	"fmt"
	"time"
)

const (
	sizeBytes = 1 << 20 // 1 MiB
	stride    = 64
)

func measure(step int) time.Duration {
	data := make([]int64, sizeBytes/8)
	start := time.Now()
	var sum int64
	for i := 0; i < len(data); i += step {
		sum += data[i]
	}
	if sum == 42 {
		fmt.Println("impossible")
	}
	return time.Since(start)
}

func main() {
	seq := measure(1)
	sparse := measure(stride)
	fmt.Printf("Sequential step=1: %v\n", seq)
	fmt.Printf("Sparse stride=%d: %v\n", stride, sparse)
}

File: lessons/l1_cache_hit_rate/l1_cache_hit_rate.py
import array
import time

SIZE_BYTES = 1 << 20
STRIDE = 64

def measure(step: int) -> int:
    data = array.array("l", [0]) * (SIZE_BYTES // 8)
    acc = 0
    start = time.perf_counter_ns()
    for i in range(0, len(data), step):
        acc += data[i]
    if acc == 42:
        print("impossible")
    return time.perf_counter_ns() - start

if __name__ == "__main__":
    print(f"Step=1   : {measure(1)} ns")
    print(f"Stride={STRIDE}: {measure(STRIDE)} ns")
